knimin/lib/data_access.py

```diff
@@ -2630,283 +2628,120 @@ def delete_study(self, study_id):
+
+            # Check if there are any samples that we need to remove
+            samples = self.get_study_samples(study_id)
             if samples:
-                samples_to_delete = [x for x in samples if not samples[x]]
-                if samples_to_delete:
-                    self.delete_samples(samples_to_delete)
-                if len(samples) > len(samples_to_delete):
-                    sql = """DELETE FROM pm.study_sample
-                             WHERE study_id = %s"""
-                    TRN.add(sql, [study_id])
-            sql = """DELETE FROM pm.study
-                     WHERE study_id = %s"""
+                sql = "DELETE FROM pm.study_sample WHERE study_id = %s"
+                TRN.add(sql, [study_id])
+                sql = "DELETE FROM pm.sample WHERE sample_id IN %s"
+                TRN.add(sql, [tuple(samples)])
+
+            # Delete the study
+            sql = "DELETE FROM pm.study WHERE study_id = %s"
             TRN.add(sql, [study_id])
             TRN.execute()
 
-    def _sample_exists(self, sample_id):
-        """Checks whether a sample ID exists
-
-        Parameters
-        ----------
-        sample_id : str
-            ID of the sample to check
-
-        Returns
-        ------
-        bool
-            Whether the sample exists
-        """
-        with TRN:
-            sql = """SELECT EXISTS (SELECT 1 FROM pm.sample
-                                    WHERE sample_id = %s)"""
-            TRN.add(sql, [sample_id])
-            return TRN.execute_fetchlast()
-
-    def _samples_exist(self, sample_ids, exist=True):
-        """Confirms that given sample ID(s) (do not) exist
-
-        Parameters
-        ----------
-        sample_ids : list of str
-            IDs of the sample to check
-        exist : bool, optional, default: True
-            True: checks if exist
-            False: checks if not exist
-
-        Raises
-        ------
-        ValueError
-            If one or more sample IDs do not exist (when exist == True)
-            If one or more sample IDs already exist (when exist == False)
-        """
-        with TRN:
-            sql = """SELECT sample_id
-                     FROM pm.sample
-                     WHERE sample_id IN %s"""
-            TRN.add(sql, [tuple(sample_ids)])
-            res = TRN.execute_fetchflatten()
-            if exist:
-                res = set(sample_ids) - set(res)
-                if res:
-                    raise ValueError('Sample ID(s) %s do not exist.' %
-                                     ', '.join(sorted(res)))
-            else:
-                if res:
-                    raise ValueError('Sample ID(s) %s already exist.' %
-                                     ', '.join(sorted(res)))
-
-    def _validate_samples(self, samples):
-        """Validate sample properties
-
-        Confirms that the barcodes and study IDs to be assigned to samples are
-        valid.
+    def set_study_samples(self, study_id, samples):
```

Comment:
I am bit unclear with the logics behind this function. What are all possible mechanism(s) of adding samples to labadmin? Here I see one mechanism: the user selects samples associated with a Qiita study in the plate mapper interface. Therefore, "create a sample" === "plate a sample".

Then, in what situations will samples be dissociated from a plate & removed? In the docstring of this function I learned that a plated sample cannot be removed.

It will be great if you can add some explanations.